LibraBFT/Abstract/BFT.agda

```diff
@@ -8,29 +8,40 @@ open import LibraBFT.Lemmas
 open import LibraBFT.Abstract.Types
 open import LibraBFT.Base.PKCS
 
-  -- This module provides a utility function to make it easy to provide the
-  -- bft-lemma for implementations that assume one peer, one vote and assume
-  -- that: at most bizF members are byzantine; authorsN ≥ suc (3 * bizF);
-  -- and a list of Members is a quorum if it contains at least authorsN ∸
-  -- bizF = QSize distinct Members. The bft-lemma is the last lemma in this
-  -- file and proves that in the intersection of any two lists with at least
-  -- QSize distinct Members there is an honest Member.
+  -- This module provides a utility function to make it easy to
+  -- provide the bft-lemma for implementations in which the
+  -- participants can hve different voting power.
+
+  -- The module should be parametrized with the number of participants
+  -- - `authorsN`, and with a function - `votPower` - that assigns to
+  -- each participant its voting power. The parameter `N` corresponds
+  -- to the total voting power of all participants, as required by the
+  -- parameter `totalVotPower` in the inner module. These
+  -- implementations should assume a fixed unknown subset of malicious
+  -- nodes - Byzantine - but should also assume a security threshold
+  -- `bizF`, such that N > 3 * bizF, which should be provided in
+  -- parameter `isBFT`.  Finally, the `bft-assumption` states that the
+  -- combined voting power of Byzantine nodes must not exceed the
+  -- security threshold `bizF`. A quorum is a subset of distinct nodes
+  -- whose combined voting power `QSize` satisfies `QSize ≥ N ∸ bizF`,
+  -- received as the parameter `quorumVotPower` in the ineer module.
+
+  -- The bft-lemma is the last lemma in this file and proves that in
+  -- the intersection of any quoruns there is an honest Member.
 
 module LibraBFT.Abstract.BFT
   (authorsN  : ℕ)
-  (bizF      : ℕ)
-  (N         : ℕ)
   (votPower  : Fin authorsN → ℕ)
+  (N         : ℕ)
+  (bizF      : ℕ)
   (isBFT     : N ≥ suc (3 * bizF))
+  (QSize     : ℕ)
```

Comment:
I don't think we should introduce this `QSize` parameter.  It'd be better to have the final property say:
```
   bft-lemma : {xs ys : List Member}
             → IsSorted _<Fin_ xs → IsSorted _<Fin_ ys
             → N ∸ bizF ≤ CombinedPower xs
             → N ∸ bizF ≤ CombinedPower ys
             → ∃[ α ] (α ∈ xs × α ∈ ys × Meta-Honest-PK (getPubKey α))
```
and eliminate both the `QSize` and `quorumVotPower` module parameters.  The implementation could use the same proof to invoke `bft-lemma` that it would use to provide the `quorumVotPower` module parameter, but this way is slightly more general, and does not require the implementation to identify a specific quorum size that is the same for all quorums (as long as it ensures that every quorum has "enough" voting power).